python/pylibraft/pylibraft/common/device_ndarray.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import numpy as np
+
+import rmm
+
+
+class device_ndarray:
+    """
+    pylibraft.device_ndarray is meant to be a very lightweight
+    __cuda_array_interface__ wrapper around a numpy.ndarray.
+    """
+
+    def __init__(self, np_ndarray):
+        """
+        Construct a pylibraft.device_ndarray wrapper around a numpy.ndarray
+
+        Parameters
+        ----------
+        ndarray : A numpy.ndarray which will be copied and moved to the device
+
+        Examples
+        --------
+        The device_ndarray is __cuda_array_interface__ compliant so it is
+        interoperable with other libraries that also support it, such as
+        CuPy and PyTorch. The following usage example demonstrates
+        converting a pylibraft.device_ndarray to a cupy.ndarray:
+
+        .. code-block:: python
+
+            import cupy as cp
+            from pylibraft import device_ndarray
+
+            raft_array = device_ndarray.empty((100, 50))
+            cupy_array = cp.asarray(raft_array)
+        """
+        self.ndarray_ = np_ndarray
+        order = "C" if self.c_contiguous else "F"
+        self.device_buffer_ = rmm.DeviceBuffer.to_device(
+            self.ndarray_.tobytes(order=order)
+        )
+
+    @staticmethod
+    def empty(shape, dtype=np.float32, order="C"):
+        """
+        Return a new device_ndarray of given shape and type, without
+        initializing entries.
+
+        Parameters
+        ----------
+        shape : int or tuple of int
+                Shape of the empty array, e.g., (2, 3) or 2.
+        dtype : data-type, optional
+                Desired output data-type for the array, e.g, numpy.int8.
+                Default is numpy.float32.
+        order : {'C', 'F'}, optional (default: 'C')
+                Whether to store multi-dimensional dat ain row-major (C-style)
+                or column-major (Fortran-style) order in memory
+        """
+        arr = np.empty(shape, dtype=dtype, order=order)
+        return device_ndarray(arr)
```

Comment:
This maybe extremely nit-picky - but I prefer using `@classmethod` instead of `@staticmethod` for alternate constructors in python, since `@classmethod` works with subclasses: 

```suggestion
    @classmethod
    def empty(cls, shape, dtype=np.float32, order="C"):
        """
        Return a new device_ndarray of given shape and type, without
        initializing entries.

        Parameters
        ----------
        shape : int or tuple of int
                Shape of the empty array, e.g., (2, 3) or 2.
        dtype : data-type, optional
                Desired output data-type for the array, e.g, numpy.int8.
                Default is numpy.float32.
        order : {'C', 'F'}, optional (default: 'C')
                Whether to store multi-dimensional dat ain row-major (C-style)
                or column-major (Fortran-style) order in memory
        """
        arr = np.empty(shape, dtype=dtype, order=order)
        return cls(arr)
```